src/wsiToDcm.cpp

```diff
@@ -129,6 +147,119 @@ int32_t get_openslide_level_for_downsample(openslide_t *osr,
     }
     return (levelToGet - 1);
 }
+
+std::unique_ptr<SlideLevelDim> get_slide_level_dim(
+                                  const bool retile,
+                                  const std::vector<int> &downsamples,
+                                  const bool preferProgressiveDownsampling,
+                                  const bool floorCorrectDownsampling,
+                                  openslide_t *osr,
+                                  const int32_t level,
+                                  const int64_t frameWidth,
+                                  const int64_t frameHeight,
+                                  SlideLevelDim *priorLevel,
+                                  DCM_Compression levelCompression,
+                                  const int64_t initialX,
+                                  const int64_t initialY,
+                                  SlideLevelDim * smallestSlideLevel) {
+  int32_t levelToGet = level;
+  int64_t downsample = 1;
+  bool readOpenslide = false;
+  if (retile) {
+    if (downsamples.size() > level && downsamples[level] >= 1) {
+      downsample = downsamples[level];
+    } else {
+      downsample = static_cast<int64_t>(pow(2, level));
+    }
+  }
+  /*
+     DICOM requires uniform pixel spacing across downsampled image
+     for pixel spacing based metrics to produce images with compatiable
+     coordinate systems across zoom levels.
+
+     Downsampled acquistions can have in image dimensions which are
+     non-interger multiples of the highest magnification. Example: Aperio svs
+     imaging, E.g. (40x -> 10x reports the 10x image has having a
+     downsampling  factor of 4.00018818010427. This results in non-uniform
+     scaling of the pixels and can result in small, but signifcant
+     mis-alignment in the downsampled imageing. Flooring, the multiplier
+     returned by openslide_get_level_downsample corrects this by restoring
+     consistent downsamping and pixel spacing across the image.
+  */
+  double multiplicator;
+  double downsampleOfLevel;
+  int64_t levelWidth;
+  int64_t levelHeight;
+  bool generateUsingOpenSlide = true;
+  // ProgressiveDownsampling
+  if (preferProgressiveDownsampling) {
+    if (priorLevel != NULL) {
+      multiplicator = static_cast<double>(priorLevel->downsample);
+      downsampleOfLevel = static_cast<double>(downsample) / multiplicator;
+      // check that downsampling is going from higher to lower magnification
+      if (downsampleOfLevel >= 1.0) {
+        levelWidth = priorLevel->levelWidthDownsampled;
+        levelHeight = priorLevel->levelHeightDownsampled;
+        generateUsingOpenSlide = false;
+      }
```

Comment:
Should we be validating somewhere that downsample ints are specified in ascending order when progressive mode is enabled? Or maybe just documenting in the comment of that flag that this code path won't run unless downsamples are specified in an increasing order